Make Key's last type parameter have a nominal role

RyanGlScott opened the pull request:
Currently, `Key a` is a simple newtype around `Unique`. What that doesn't convey is an invariant that `a` must correspond to the type of the data in a `Vault`, or bad things can happen:

```haskell
-- Vault.hs
module Main (main) where

import           Data.Coerce (coerce)
import qualified Data.Vault.Lazy as Vault

main :: IO ()
main = do
  ki <- Vault.newKey :: IO (Vault.Key Int)
  ko <- Vault.newKey :: IO (Vault.Key Ordering)
  let v1  = Vault.insert ki 42 Vault.empty
      v2  = Vault.insert ko EQ v1
      ki2 = coerce ko :: Vault.Key Int
  print $ Vault.lookup ki  v2
  print $ Vault.lookup ko  v2
  print $ Vault.lookup ki2 v2
```

```bash
$ runhaskell Vault.hs
Just 42
Just EQ
Just (-5188146770730811392)
```

We can freely `coerce` between `Key`s of different types because GHC infers the type parameter in `Key a` to be phantom. To prevent this, this pull request forces `a` to have a nominal role so that you can't `coerce` your way into undefined behavior like the above example.

HeinrichApfelmus commented on src/Data/Vault/ST/backends/GHC.hs:
```diff
@@ -18,6 +18,10 @@ fromAny = unsafeCoerce
 newtype Key s a = Key Unique
 
+#if __GLASGOW_HASKELL__ >= 708
+type role Key phantom nominal
```
I'm not completely sure about the status of the parameter `s` in the `ST` monad, but I think it should be `nominal` as well. Otherwise, it would be possible to write bad programs using `coerce` as well. One might actually want to report this issue to GHC HQ.

HeinrichApfelmus commented on vault.cabal:
```diff
@@ -45,6 +45,8 @@ Library
 
     extensions:         CPP
+    if flag(UseGHC) && impl(ghc >= 7.8)
+      extensions:       RoleAnnotations
```
Could you move the `RoleAnnotations` language extension from the cabal file to the source file by using a `LANGUAGE` pragma?

HeinrichApfelmus commented:
I knew nothing about the type roles extension, but apparently, it can still break my code. :smile: 

I have added some comments to the diff. I'm happy to merge if they are addressed.

RyanGlScott commented:
I've addressed your comments with another commit.

> but apparently, it can still break my code.

I should probably be more precise in describing what the "issue" is. There's nothing inherently _wrong_ with your code as is. In GHC 7.8, [type roles](https://ghc.haskell.org/trac/ghc/wiki/Roles) were introduced to support a notion of type-safe coercions. The tl;dr version is that.

* Nominal equality is type equality as we know it. `Int` is nominally equal to `Int`. `Int` isn't nominally equal to `Char`.
* Representational equality is more permissive. A `newtype` is representationally equal to its underlying type. For example, if you have `newtype WrappedInt = WrapInt Int`, then `Int` is representationally equal to `Int` and `WrappedInt`, but not `Char`.
* Phantom equality is even more permissive. Any two types are phantom equal, so `Int` is phantom equal to `Int`, `WrappedInt`, `Char`, or anything else.

`Data.Coerce.coerce` is thus "safer" than `unsafeCoerce` because it will only coerce between two types that are representationally equal, making things like `coerce 'a' :: Int` fail to typecheck.

The type role machinery is imperfect, however, because it has to guess what the roles of type parameters are, and sometimes its guesses are too permissive. For example, if you have

```haskell
newtype ID a = ID String
```

Then since `a` appears nowhere in the constructor, GHC infers `a`'s role to be phantom. Usually, that's what you want, since the type of `a` doesn't affect an `Id`'s internal structure at all, making it safe to `coerce` between, say, `ID Int` and `ID Char`.

There are some corner cases, however (`vault` being one of them), where you don't want a phantom type parameter to actually have a phantom type _role_, since the phantom type might actually be encoding an invariant about the data's structure. In that case, `coerce` might break that invariant, so you can use `RoleAnnotations` to rule this out.

(You can still `unsafeCoerce` to get a `Key Int` from a `Key Char`, but that's to be expected from something marked "unsafe") :)

HeinrichApfelmus commented:
Well, in a way `coerce` does expose internal structure that was not exposed before — unless we prevent it with `nominal` type roles. I guess is just means that `coerce` is "semi-unsafe".

Anyway, thanks for the patch! I'll upload a new minor version to hackage.
